### PR TITLE
Added support for Promise class properties

### PR DESCRIPTION
**Pull request by derekxu16**
Emitted extension getters and setters to expose them as Futures in the generated facade

### CODE REVIEW

File: lib/declaration.ts [hunk 1]
```diff
@@ -979,32 +1002,44 @@ export default class DeclarationTranspiler extends base.TranspilerBase {
...
-      if (ts.isTypeReferenceNode(declaration.type)) {
-        this.maybeVisitTypeArguments(declaration.type);
+      if (ts.isPropertyDeclaration(declaration) || ts.isPropertySignature(declaration)) {
+        visitTypeAndArguments(declaration.type);
+        this.emit('get');
+        this.visitName(declaration.name);
+        this.emit('{');
+        castThisToPrivateClass();
+        this.emitExtensionGetterBody(declaration);
+        this.emit('}');
```
Comment (derekxu16) [inline lib/declaration.ts]: Should I move `this.emit('{')`, `this.emit('}')` and `castToPrivateClass` inside emitExtensionBody?

**Comment by jacob314** [comment]
@ditman can you take a look?

File: lib/declaration.ts [hunk 2]
```diff
@@ -529,7 +530,11 @@ export default class DeclarationTranspiler extends base.TranspilerBase {
         this.visitMergingOverloads(sourceFile.statements);
         if (this.containsPromises) {
           this.addImport('package:js/js_util.dart', 'promiseToFuture');
-          this.emit(`@JS() abstract class Promise<T> {}\n`);
+          this.emit(`@JS()
+                     abstract class Promise<T> {
```
Comment (jacob314) [inline lib/declaration.ts]: add a TODO to put this in a package instead of emitting it each time.

File: lib/declaration.ts [hunk 3]
```diff
@@ -963,14 +968,32 @@ export default class DeclarationTranspiler extends base.TranspilerBase {
...
+        this.emit('Future');
+        if (ts.isTypeReferenceNode(type)) {
+          this.maybeVisitTypeArguments(type);
+        }
+      } else {
+        this.fc.visit(type);
+      }
+    };
+
+    const castThisToPrivateClass = () => {
```
Comment (jacob314) [inline lib/declaration.ts]: add the word emit to this method as it is emitting code.

Comment (jacob314) [inline lib/declaration.ts]: why isn't this a method? based on the usage it can just be a method on this class.

File: lib/declaration.ts [hunk 4]
```diff
@@ -979,32 +1002,44 @@ export default class DeclarationTranspiler extends base.TranspilerBase {
...
-        this.maybeVisitTypeArguments(declaration.type);
+      if (ts.isPropertyDeclaration(declaration) || ts.isPropertySignature(declaration)) {
+        visitTypeAndArguments(declaration.type);
+        this.emit('get');
+        this.visitName(declaration.name);
+        this.emit('{');
+        castThisToPrivateClass();
+        this.emitExtensionGetterBody(declaration);
+        this.emit('}');
+        if (!base.isReadonly(declaration)) {
```
Comment (jacob314) [inline lib/declaration.ts]: seems like this code is almost completely duplicated . Can you tweak so more code is reused?

File: lib/declaration.ts [hunk 5]
```diff
@@ -1031,15 +1066,31 @@ export default class DeclarationTranspiler extends base.TranspilerBase {
...
+        this.emit('return promiseToFuture(tt.');
+        this.visitName(mergedDeclaration.name);
         this.visitParameters(ts.createNodeArray(mergedDeclaration.parameters), {namesOnly: true});
         this.emit(');\n');
       }
     }
   }
+
+  private emitExtensionGetterBody(declaration: ts.PropertyDeclaration|ts.PropertySignature) {
+    this.emit('return promiseToFuture(tt.');
```
Comment (jacob314) [inline lib/declaration.ts]: you should make `tt` a constant and replace all usages with the constant so it is easy to tweak `tt` to something else in the future

**Comment by ditman** [comment]
> looks reasonable overall. @ditman will take a look at the cl in a bit more detail.

Taking a look, it's the first time I see this package so let me read some docs, and code around this as well!

**Comment by ditman** [comment]
As a new comer to this it'd help if you added a test to this feature, a simple "happy case" would be probably enough (but there's quite a bit of branching).

Something like: 

This is the incoming TS code --> This is the expected Dart output

Is it doable to add a unit test to this? Other than that, LGTM!

**Comment by derekxu16** [comment]
> Is it doable to add a unit test to this? Other than that, LGTM!

I added some tests for this behavior in test/declaration_test.ts. I think you missed it because I edited that file in this pr's first commit, but not the second commit.

**Comment by ditman** [comment]
> > Is it doable to add a unit test to this? Other than that, LGTM!
> 
> I added some tests for this behavior in test/declaration_test.ts. I think you missed it because I edited that file in this pr's first commit, but not the second commit.

🤦‍♂ totally, my bad!

File: lib/declaration.ts [hunk 6]
```diff
@@ -540,26 +557,17 @@ export default class DeclarationTranspiler extends base.TranspilerBase {
...
         if (!this.hasNodeFlag(varDecl, ts.NodeFlags.Const)) {
-          this.emitNoSpace(';');
-          this.maybeEmitJsAnnotation(varDecl);
-          this.emit('external');
-          this.emit('set');
-          this.visitName(varDecl.name);
-          this.emitNoSpace('(');
-          this.visit(varDecl.type);
-          this.emit('v)');
+          this.emitGetterOrSetter(
```
Comment (jacob314) [inline lib/declaration.ts]: nit: rename emitGetterOrSetter to emitProperty

File: lib/declaration.ts [hunk 7]
```diff
@@ -815,30 +822,18 @@ export default class DeclarationTranspiler extends base.TranspilerBase {
...
-        if (isStatic) this.emit('static');
-        this.emit('set');
-        this.visitName(decl.name);
-        this.emitNoSpace('(');
-        this.visit(decl.type);
-        this.emit('v');
-        this.emitNoSpace(')');
-        this.emitNoSpace(';');
+        this.emitGetterOrSetter(
+            {mode: 'setter', declaration: decl, emitJsAnnotation: false, isExternal: true});
```
Comment (jacob314) [inline lib/declaration.ts]: make 'mode' an enum rather than a string.

File: lib/declaration.ts [hunk 8]
```diff
@@ -963,48 +958,106 @@ export default class DeclarationTranspiler extends base.TranspilerBase {
     this.emitNoSpace(';');
   }
 
-  private emitMethodsAsExtensions(
-      className: ts.Identifier, visitName: () => void, visitNameOfExtensions: () => void,
+  private emitGetterOrSetter({mode, declaration, emitJsAnnotation, isExternal, emitBody}:
```
Comment (jacob314) [inline lib/declaration.ts]: provide types for these arguments.

Comment (derekxu16) [reply lib/declaration.ts]: The first part is just specifying the names for object destructuring the types are defined by `:emitGetterOrSetterOptions`

**Comment by jacob314** [comment]
lgtm

**Pull request merged**